File: src/main/java/com/chatchatabc/jpademojava/domain/repository/UserRepository.java
package com.chatchatabc.jpademojava.domain.repository;

import com.chatchatabc.jpademojava.domain.model.Country;
import com.chatchatabc.jpademojava.domain.model.User;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
public interface UserRepository extends JpaRepository<User, String> {

    /**
     * Find user by username
     *
     * @param username
     * @return
     */
    Optional<User> findByUsername(String username);

    /**
     * Find users by country
     *
     * @param country
     * @param pageable
     * @return
     */
    Page<User> findUsersByCountry(Country country, Pageable pageable);
}

File: src/main/kotlin/com/chatchatabc/jpademo/domain/repository/UserRepository.kt
package com.chatchatabc.jpademo.domain.repository

import com.chatchatabc.jpademo.domain.model.Country
import com.chatchatabc.jpademo.domain.model.User
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Repository
import java.util.*

@Repository
interface UserRepository : JpaRepository<User, String> {

    /**
     * Find user by username
     */
    fun findByUsername(username: String): Optional<User>

    /**
     * Find user by country
     */
    fun findUsersByCountry(country: Country, pageable: Pageable): Page<User>
}